src/main/java/org/cryptacular/CiphertextHeaderFactory.java

```diff
@@ -0,0 +1,344 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.cryptacular;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import javax.crypto.SecretKey;
+import org.cryptacular.util.ByteUtil;
+
+/**
+ * Static factory class that produces the correct instance of {@link CiphertextHeader} given a byte array or byte
+ * stream that presumably contains an encoded ciphertext header.
+ *
+ * @author Middleware Services
+ */
+public final class CiphertextHeaderFactory
+{
+  /** Prevent instantiating this class. */
+  private CiphertextHeaderFactory() {}
+
+
+  /**
+   * Decodes the ciphertext header at the start of the given byte array. Supports both v1 and v2 formats.
+   *
+   * @param  data  Ciphertext data with prepended header.
+   * @param  keyLookup  Decryption key lookup function.
+   *
+   * @return  Ciphertext header instance.
+   */
+  public static CiphertextHeader decode(final byte[] data, final Function<String, SecretKey> keyLookup)
```

Comment:
Note that I could have gone further here in trying to future-proof the API by defining a `DecodeParams` type as the second argument that contains anything that might ever be required to decode a header. Problem was that I sincerely couldn't think of anything other than a secret key used for an HMAC that might ever be used. If you can imagine something else, that might be reason to insulate details in a domain class.

Reply:
Nothing comes to mind.